### DISCUSSION TITLE
Phantom collision due to convex manifold depth being too large

### DISCUSSION THREAD

**Discussion by Grimelios** [bepu/bepuphysics2#267] [Q&A] [answered]
Hello! I've run into a scenario where Bepu is reporting a collision (i.e. triggering a narrow phase callback with a manifold with contacts with positive depth), but the depth on each contact is larger than should be possible. Specifically, an upright, kinematic cylinder (representing a video game character) is having a collision reported against a vertical triangle within a mesh, i.e. a triangle with a unit-Z normal. From a top-down perspective, the scenario looks like this: [https://imgur.com/a/QQnhopl](https://imgur.com/a/QQnhopl) (I've included the triangle's normal in this image, although it may not be relevant to my question here).

As shown, a manifold with positive depth is generated by Bepu despite the cylinder not touching the triangle. In other words, **the reported depth is (significantly) greater than the cylinder's radius**.

In this specific instance, I can work around the problem by manually comparing contact depth vs. cylinder radius, but it's still concerning to me that my game might respond to other "phantom" collisions. Feels like an edge case (potentially resulting in a game crash) waiting to happen. Hoping for some clarity on whether this scenario is considered "valid" and, if so, how I might identify (and skip processing on) such collisions in the general case.

Thank you!

**Comment by RossNordby** [comment]
It definitely doesn't sound valid! Is this on latest master? Could you provide a minimal reproduction in the demos?

**Reply by Grimelios** [reply]
Appreciate the quick response! I'm currently using Bepu 2.4.0 as a NuGet package. I haven't tested against the latest master branch, although I'll plan to try that first. If the problem persists, I'll look into creating a reproducible example this evening and get back to you.

**Comment by Grimelios** [comment, marked as answer]
Welp, false alarm. Turns out it's a problem with my game logic, not Bepu.

I'm using an `INarrowPhaseCallbacks` implementation that allows objects to track specific collisions (using a function with the signature `OnContact(CollidableReference collidable, int childIndex)`). The mistake I made is that `INarrowPhaseCallbacks` has two overloads each for `AllowContactGeneration` and `ConfigureContactManifold`, one that takes child indexes and one that doesn't. For a body with multiple collidables (like a `Mesh`), _both_ functions are called while processing collisions, not _just_ the child version. My mistake, then, was calling `OnContact(collidable, 0)` (note the hardcoded `0`) for the non-child overload. By a hilarious and fortunate coincidence, the 0th triangle in my current mesh happens to be a wall (the vertical triangle mentioned in my original question), hence a seemingly phantom collision.

To summarize, **Bepu is reporting collisions accurately, but my game logic erroneously pulled the wrong triangle**. I apologize for not verifying more thoroughly before posting. Ross, would you prefer I delete this post, or leave it up and mark my own answer as the "solution"?

**Reply by RossNordby** [reply]
> Ross, would you prefer I delete this post, or leave it up and mark my own answer as the "solution"?

Mild preference for leaving it up and marking the solution, I suppose- having more words for people to search for tends to be helpful. Who knows, maybe someone else will do the same thing :P

**Reply by Grimelios** [reply]
Will do! Thanks again.
